executor/src/witgen/jit/single_step_processor.rs

```diff
@@ -31,64 +31,129 @@ impl<'a, T: FieldElement> SingleStepProcessor<'a, T> {
         &self,
         can_process: CanProcess,
     ) -> Result<Vec<Effect<T, Variable>>, String> {
-        // All witness columns in row 0 are known.
+        self.generate_code_for_branch(can_process, self.initialize_witgen(), Default::default())
+    }
+
+    pub fn generate_code_for_branch<CanProcess: CanProcessCall<T> + Clone>(
+        &self,
+        can_process: CanProcess,
+        mut witgen: WitgenInference<'a, T, NoEval>,
+        mut complete: HashSet<u64>,
+    ) -> Result<Vec<Effect<T, Variable>>, String> {
+        self.process_until_no_progress(can_process.clone(), &mut witgen, &mut complete);
+
+        // Check that we could derive all witness values in the next row.
+        let unknown_witnesses = self
+            .unknown_witness_cols_on_next_row(&witgen)
+            .sorted()
+            .collect_vec();
+
+        let missing_identities = self.machine_parts.identities.len() - complete.len();
+        let code = if unknown_witnesses.is_empty() && missing_identities == 0 {
+            witgen.code()
+        } else {
+            let Some((most_constrained_var, _)) = witgen
+                .known_variables()
+                .iter()
+                .filter_map(|var| witgen.range_constraint(var).map(|rc| (var, rc)))
+                .filter(|(_, rc)| rc.try_to_single_value().is_none())
+                .sorted()
+                .min_by_key(|(_, rc)| rc.range_width())
+            else {
+                let incomplete_identities = self
+                    .machine_parts
+                    .identities
+                    .iter()
+                    .filter(|id| !complete.contains(&id.id()));
+                return Err(format!(
+                    "Unable to derive algorithm to compute values for witness columns in the next row and\n\
+                    unable to branch on a variable. The following columns are still missing:\n{}\nThe following identities have not been fully processed:\n{}",
+                    unknown_witnesses.iter().map(|wit| self.fixed_data.column_name(wit)).format(", "),
+                    incomplete_identities.map(|id| format!("    {id}")).join("\n")
+                ));
+            };
+
+            let (common_code, condition, other_branch) =
+                witgen.branch_on(&most_constrained_var.clone());
+
+            // TODO Tuning: If this fails (or also if it does not generate progress right away),
+            // we could also choose a different variable to branch on.
+            let left_branch_code =
+                self.generate_code_for_branch(can_process.clone(), witgen, complete.clone())?;
+            let right_branch_code =
+                self.generate_code_for_branch(can_process, other_branch, complete)?;
+            if left_branch_code == right_branch_code {
```

Comment:
In which cases would this happen?

Reply:
It happens when the chosen variable is not helpful. This is the case already in the test here.